### PR TITLE
Upgrade to Lucene 9.9.1

### PR DESCRIPTION
**Pull request by lintool**
WIP, not ready for review.
All Unit tests pass.

@ChrisHegarty @tteofili @jpountz

### CODE REVIEW

**Comment by ChrisHegarty** [comment]
This looks fine to me.   Just a heads up, we're in the process of releasing a Lucene 9.9.1 to fix a couple of critical issues. If it suits we can delay this PR a little (to say early next week) so as to pickup 9.9.1. Or merge it and bump to 9.9.1 when it is available.

**Comment by lintool** [comment]
> This looks fine to me. Just a heads up, we're in the process of releasing a Lucene 9.9.1 to fix a couple of critical issues. If it suits we can delay this PR a little (to say early next week) so as to pickup 9.9.1. Or merge it and bump to 9.9.1 when it is available.

Sounds good! I need to run a bunch of tests anyway!

**Comment by lintool** [comment]
Hi @ChrisHegarty can you check if I'm using `Lucene99HnswScalarQuantizedVectorsFormat` correctly?

Here are my preliminary experiments on cosDPR-distil on MS MARCO passage dev -

Default (fp32):

```
# Total 8,841,823 documents indexed in 00:54:47
# 6980 queries processed in 00:02:57 = ~39.42 q/s
# 26G	indexes/lucene-hnsw.msmarco-passage-cos-dpr-distil
```

Quantized (int8):

```
# Total 8,841,823 documents indexed in 00:45:26
# 6980 queries processed in 00:01:41 = ~68.59 q/s
# 33G	lucene-hnsw.msmarco-passage-cos-dpr-distil
```

QPS increased, but index got _bigger_?!

Not sure if it matters, but I'm giving `iwc.setRAMBufferSizeMB` a generous 64GB buffer. This is on my Mac Studio with M1 Ultra processor.

**Comment by ChrisHegarty** [comment]
@lintool Do your benchmarks use a recent JDK (>= JDK 20)? If so, can you confirm that the JDK Panama Vector API module is present at runtime (`--add-modules=jdk.incubator.vector`). Lucene will use it if present, to improve vector distance computations.  On an M1, it'll use 128 bit Neon instructions for floating point arithmetic, while AVX 2/512 on x64.

**Comment by lintool** [comment]
> @lintool Do your benchmarks use a recent JDK (>= JDK 20)? If so, can you confirm that the JDK Panama Vector API module is present at runtime (`--add-modules=jdk.incubator.vector`). Lucene will use it if present, to improve the speed of vector distance computations. On an M1, it'll use 128 bit Neon instructions for floating point arithmetic, while AVX 2/512 on x64.

Nope, this is still JDK 11. Will queue this up as something to look at!

**Comment by lintool** [comment]
hey @jpountz @ChrisHegarty for HNSW indexing, how do I set the index writer config to generate larger segments? I'm setting `config.setRAMBufferSizeMB` to 64GB, but looking at the index, I only see segments of 1.5GB.

**Comment by jpountz** [comment]
You can't, Lucene has a per-segment limit of ~2GB in memory (which then usually translates into a smaller size on disk). https://lucene.apache.org/core/9_9_0/core/org/apache/lucene/index/IndexWriterConfig.html#setRAMPerThreadHardLimitMB(int)

**Comment by lintool** [comment]
> You can't, Lucene has a per-segment limit of ~2GB in memory (which then usually translates into a smaller size on disk). https://lucene.apache.org/core/9_9_0/core/org/apache/lucene/index/IndexWriterConfig.html#setRAMPerThreadHardLimitMB(int)

I see. Thanks for the quick response!

**Comment by lintool** [comment]
FYI @jpountz @ChrisHegarty more thorough experiments documented in #2292

File: src/main/java/io/anserini/index/IndexCollection.java [hunk 1]
```diff
@@ -190,7 +177,7 @@ public IndexCollection(Args args) throws Exception {
     }
 
     final Directory dir = FSDirectory.open(Paths.get(args.index));
-    final IndexWriterConfig config = new IndexWriterConfig(getAnalyzer());
+    final IndexWriterConfig config = new IndexWriterConfig(getAnalyzer()).setCodec(new Lucene99Codec());
```
Comment (tteofili) [inline src/main/java/io/anserini/index/IndexCollection.java:180]: `lucene99` is the default in Lucene 9.9, so I guess it is fine not to specify it manually.

File: src/main/java/io/anserini/index/IndexInvertedDenseVectors.java [hunk 2]
```diff
@@ -104,7 +100,7 @@ public IndexInvertedDenseVectors(Args args) {
 
     try {
       final Directory dir = FSDirectory.open(Paths.get(args.index));
-      final IndexWriterConfig config = new IndexWriterConfig(analyzer).setCodec(new Lucene95Codec());
+      final IndexWriterConfig config = new IndexWriterConfig(analyzer).setCodec(new Lucene99Codec());
```
Comment (tteofili) [inline src/main/java/io/anserini/index/IndexInvertedDenseVectors.java:103]: same as above, we can probably avoid setting the codec manually.

**Review by tteofili — APPROVED** [review]
LGTM

**Pull request merged**